### DISCUSSION TITLE
Can I return a partial success if one element of an array is invalid?

### DISCUSSION THREAD

**Discussion by camjackson** [apollographql/apollo-server#5889] [Q&A] [answered]
Say I have a query that returns an array of `Customer` objects, each one having an `id`, `name`, and `email`, all of which are non-nullable. My resolver loads the data from some source, maybe a database or maybe a downstream system. Most of the data that comes back is fine, but maybe for some reason we're missing the email address for a single customer.

Currently if my resolver just returns the array of all known customers, including the dodgy one, then the entire operation will fail with something like:

```json
{
  "data": {
    "customers": null
  },
  "errors": [{
    "message": "Cannot return null for non-nullable field Customer.email"
    // More error stuff here
  }]
}
```

I get that the validation is supposed to be strict, but now on my UI it fails to load the entire list of customers just because one of them is invalid. It makes it too easy for a single record to poison the entire database.

What I would like is a way to still return the error, but to also have the `data` field be populated with all of the other, valid `Customer` records. Is this possible to do with apollo server? Or do I need to manually validate all of the data before trying to return it from my resolver? That would be quite arduous as my real data structure is much more complex.

**Comment by glasser** [comment, marked as answer]
Null propagation through non-null types is part of the GraphQL spec.

Would you like to end up with `null` as the bad customer's email?  In that case, declare the `email` field as nullable (eg `String` instead of `String!`).

Would you like the bad customer's entry in the list to end up `null` (alongside other entries which are non-null objects)? In that case, declare the `customers` field as `[Customer]` (or perhaps `[Customer]!`) rather than as `[Customer!]`.

**Reply by camjackson** [reply]
I was thinking that I would like for the customer with the missing data to just be removed from the result entirely. So `data` would contain the list of valid customers, and then `errors` can report any customers that can't be loaded because they have invalid data.

However, your second solution sounds pretty decent as well. If I understand correctly, if I make `Customer` itself nullable, then any field-level errors will propagate back up to the top-level array, resulting in e.g. `[customer1, null, customer3]`. Then it's up to the client to decide what to do with those, either skipping over null'd customers, or showing them in some other way. Is that right?

**Reply by glasser** [reply]
Yes.

**Reply by camjackson** [reply]
Thank you! That's very helpful!

**Reply by camjackson** [reply]
For any future readers - for this to work you also need to set `errorPolicy: 'all'` on your apollo-client query. By default, apollo-client will throw away any returned data if there is any error.
